packages/core-app-api/src/app/PluginDecoration.ts

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Minimatch, IMinimatch } from 'minimatch';
+
+import { AppOptions, CompatiblePlugin } from './types';
+
+export type PluginDecorationOptions = Pick<
+  AppOptions,
+  'pluginOwners' | 'pluginInfoDecorator'
+>;
+
+export class PluginDecoration {
+  private readonly matchers: ReadonlyArray<readonly [IMinimatch, string]>;
+  private knownPluginIds = new Set<string>();
+
+  constructor(private options: PluginDecorationOptions) {
+    const { pluginOwners = [] } = this.options;
+
+    this.matchers = pluginOwners.flatMap(rec =>
+      Object.entries(rec).map(
+        ([pkgNamePattern, owner]) =>
+          [new Minimatch(pkgNamePattern), owner] as const,
+      ),
+    );
+  }
+
+  public matchOwner = (pkgName: string): string | undefined => {
+    const match = this.matchers.find(([matcher]) => matcher.match(pkgName));
+    return match?.[1];
+  };
+
+  public decoratePlugin = <Plugin extends CompatiblePlugin>(plugin: Plugin) => {
```

Comment:
Tempted to structure the plugin info API layer in a way that we don't need this direct mutation here, but I think we can leave that a bit and focus on the other bits first

Reply:
Yeah I didn't really know what to do tbh regarding this compatible-thingie (remove the compatibility altogether after 1.0?)